### ISSUE TITLE
Enhancement: Support Parameters on All Application Layers

### ISSUE THREAD

**Issue by Goldziher** [starlite-api/starlite#187]
We could make controllers more useful by allowing users to declare shared parameters on the controller level, making it redundant to declare them per route. This will also enhance the usefulness of composition and inheritence.

Suggested implementation:

```python
from starlite import Controller, Parameter, get


class MyController(Controller):
    path = "/some-path"
    parameters = {
        "controllerHeader": Parameter(header="controller-header"),
        "controllerCookie": Parameter(header="controller-cookie"),
        "controllerQuery": Parameter(query="controller-query"),
    }

    @get("/sub-path")
    def my_route_handler(self, controllerHeader: int, controllerCookie: str, controllerQuery: list[str]) -> None:
        return None
```

As seen above, the suggestion I put forward is to use a dictionary that maps parameters keys to `Parameter` type values. The `Parameter` instances define the required data about the actual parameter, and this way of mapping it will allow us to still properly type the variables and use pydantic inference.

**Comment by peterschutt** [comment]
I can see an argument for making these another thing that we resolve all the way up to the application level. One might want to have a consistent required param across handlers registered to multiple controllers, for example.

**Comment by peterschutt** [comment]
I'm keen to take this one on.

Any thoughts on behaviour if a parameter isn't used in a handler in the scope? Config error?

What if a handler had `**kwargs` defined, should we inject it?

**Comment by Goldziher** [comment]
> I'm keen to take this one on.
> 
> Any thoughts on behaviour if a parameter isn't used in a handler in the scope? Config error?
> 
> What if a handler had `**kwargs` defined, should we inject it?

Go for it. 

You mean if no handler uses the parameter at all? 

Maybe make a new error called `UnusedParameterError`?

Currently there is no **kwargs defined, and we shouldnt add it imo. We won't be able to generate OA docs. 

On the topic of OA - you'll need to rework some stuff for this to work with the new pattern.

**Comment by Goldziher** [comment]
Good points. 

Yes, we cannot validate that a parameter is unused if we need to use it for middlewares. And since middlewares are now also on the handler level, this makes validation either impossible - or it needs to be somehow made optional.

**Comment by peterschutt** [comment]
Forgetting about the middleware thing for a second, I still like the overall idea of not needing to repeat `Parameter(...)` definitions for every handler - although its a problem I've mostly worked around it in other ways.

**Comment by Goldziher** [comment]
So middlewares are a problem we need to tackle separately I think. I'll make a proposal. 

So what's the design you want to go for?

**Comment by peterschutt** [comment]
I have stuck some thoughts in the discord dev channel, but the crux is this:

```python
@runtime_checkable
class MiddlewareProtocol(Protocol):
    parameters: dict[str, tuple] = {"middleware_required_header": (int, Parameter(header="required-header"))}

    def __init__(self, app: ASGIApp):  # pragma: no cover
        ...

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:  # pragma: no cover
        if self.validated_paraeters.get("middleware_required_header") != "13":
            ...
        else:
            ...
    
    @property
    def validated_parameters(self) -> dict[str, Any]:
        # this mapping actually constructed dynamically
        return {"middleware_required_header": 13}
```

**Comment by Goldziher** [comment]
Released in v1.6.0
